subsys/settings/src/settings_nvs.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2019 Laczen
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "settings/settings.h"
+#include "settings/settings_nvs.h"
+#include "settings_priv.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(settings, CONFIG_SETTINGS_LOG_LEVEL);
+
+struct settings_nvs_read_fn_arg {
+	struct nvs_fs *fs;
+	u16_t id;
+};
+
+static int settings_nvs_load(struct settings_store *cs, const char *subtree);
+static int settings_nvs_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_nvs_itf = {
+	.csi_load = settings_nvs_load,
+	.csi_save = settings_nvs_save,
+};
+
+static ssize_t settings_nvs_read_fn(void *back_end, void *data, size_t len)
+{
+	struct settings_nvs_read_fn_arg *rd_fn_arg;
+
+	rd_fn_arg = (struct settings_nvs_read_fn_arg *)back_end;
+
+	return nvs_read(rd_fn_arg->fs, rd_fn_arg->id, data, len);
+}
+
+int settings_nvs_src(struct settings_nvs *cf)
+{
+	cf->cf_store.cs_itf = &settings_nvs_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_nvs_dst(struct settings_nvs *cf)
+{
+	cf->cf_store.cs_itf = &settings_nvs_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_nvs_load(struct settings_store *cs, const char *subtree)
+{
+	struct settings_nvs *cf = (struct settings_nvs *)cs;
+	struct settings_nvs_read_fn_arg read_fn_arg;
+	struct settings_handler *ch;
+	char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	char buf;
+	const char *name_argv;
+	ssize_t rc1, rc2;
+	u16_t name_id = NVS_NAMECNT_ID;
+
+	name_id = cf->last_name_id + 1;
+
+	while (1) {
+
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+
+		/* In the NVS backend, each setting item is stored in two NVS
+		 * entries one for the setting's name and one with the
+		 * setting's value.
+		 */
+		rc1 = nvs_read(&cf->cf_nvs, name_id, &name, sizeof(name));
+		rc2 = nvs_read(&cf->cf_nvs, name_id + NVS_NAME_ID_OFFSET,
+			       &buf, sizeof(buf));
+
+		if ((rc1 <= 0) && (rc2 <= 0)) {
+			continue;
+		}
+
+		if ((rc1 <= 0) || (rc2 <= 0)) {
```

Comment:
Have you considered moving this check at initialization time, so that when loading a setting we can skip the second read (for the setting's value) if the setting's name is not found?

Reply:
> Have you considered moving this check at initialization time, so that when loading a setting we can skip the second read (for the setting's value) if the setting's name is not found?

No, I haven't, but it would make settings_load a bit faster. However, load function is usually called only once, so I wonder if this is worth doing.

> Generally, when could that situation happen?

Well, this is not so common, but it could happen when the device is reset during a write operation of settings value.

@nvlsianpu, @Laczen, any thoughts about this?
I would leave this as it is now (safer & slower). We can always make optimizations afterwards if somebody starts complaining about NVS performance.

Reply:
I am fine with addressing with afterwards if necessary.